presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestRecordingHiveMetastore.java

```diff
@@ -0,0 +1,579 @@
+                return Optional.of(ImmutableList.of("value"));
+            }
+
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<List<String>> getPartitionNamesByParts(String databaseName, String tableName, List<String> parts)
+        {
+            if (databaseName.equals("database") && tableName.equals("table") && parts.equals(ImmutableList.of("value"))) {
+                return Optional.of(ImmutableList.of("value"));
+            }
+
+            return Optional.empty();
+        }
+
+        @Override
+        public Map<String, Optional<Partition>> getPartitionsByNames(String databaseName, String tableName, List<String> partitionNames)
+        {
+            if (databaseName.equals("database") && tableName.equals("table") && partitionNames.contains("value")) {
+                return ImmutableMap.of("value", Optional.of(PARTITION));
+            }
+
+            return ImmutableMap.of();
+        }
+
+        @Override
+        public void addPartitions(String databaseName, String tableName, List<PartitionWithStatistics> partitions)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void dropPartition(String databaseName, String tableName, List<String> parts, boolean deleteData)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void alterPartition(String databaseName, String tableName, PartitionWithStatistics partition)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Set<String> getRoles(String user)
+        {
+            return ImmutableSet.of("role1", "role2");
+        }
+
+        @Override
+        public Set<HivePrivilegeInfo> getDatabasePrivileges(String user, String databaseName)
+        {
+            if (user.equals("user") && databaseName.equals("database")) {
+                return ImmutableSet.of(PRIVILEGE_INFO);
+            }
+
+            return ImmutableSet.of();
+        }
+
+        @Override
+        public Set<HivePrivilegeInfo> getTablePrivileges(String user, String databaseName, String tableName)
+        {
+            if (user.equals("user") && databaseName.equals("database") && tableName.equals("table")) {
+                return ImmutableSet.of(PRIVILEGE_INFO);
+            }
+
+            return ImmutableSet.of();
+        }
+
+        @Override
+        public void grantTablePrivileges(String databaseName, String tableName, String grantee, Set<HivePrivilegeInfo> privileges)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void revokeTablePrivileges(String databaseName, String tableName, String grantee, Set<HivePrivilegeInfo> privileges)
+        {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class UnimplementedHiveMetastore
```

Comment:
Might be nicer to make this a top level class since it's generic and not tied to this test